File: mid/leetCode17.py
#17. Letter Combinations of a Phone Number
def letterCombinations(digits):
    charMap = ["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
    res = []
    if not digits: return res

    def backTracking(index,block):

        if index == len(digits):
            return res.append(block)

        digit = int(digits[index])
        letters = charMap[digit]
        
        for i in range(len(letters)):
            block += letters[i]
            backTracking(index + 1, block)
            block = block[:-1]

    backTracking(0,"")
    print(res)
    return res

def letterCombinations2(digits):
    result = []
    charMap = ["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
    if not digits: return result 

    def backTrack(index,s):

        if index == len(digits):
            
            return result.append(s)

        digit = int(digits[index])
        letters = charMap[digit]

        for i in range(len(letters)):

            s += letters[i]
            backTrack(index + 1,s)
            s = s[:-1]


    backTrack(0,"")
    print(result)
    return result 

letterCombinations2("23") 

File: problems_java/leetCode17.java
package problems_java;

import java.util.ArrayList;

public class leetCode17 {

    static ArrayList<String> list = new ArrayList<String>();
    static StringBuilder temp = new StringBuilder();

    public static ArrayList<String> letterCombinations(String digits){
        if(digits == null || digits.length() == 0){
            return list;
        }
        String[] numString = {"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        backTracking(digits,numString,0);
        return list;
    }
    public static void backTracking(String digits, String[] numString, int startIndex){
        if(startIndex == digits.length()){
            list.add(temp.toString());
            return;
        }
        String str = numString[digits.charAt(startIndex) - '0'];
        for ( int i = 0 ; i < str.length(); i ++){
                temp.append(str.charAt(i));
                backTracking(digits, numString, startIndex + 1);
                temp.deleteCharAt(temp.length() - 1);
        }

    }
    public static void main(String[] args){

    }
    
}
